misoc/cores/icap.py

```diff
@@ -0,0 +1,152 @@
+from migen import *
+
+from misoc.interconnect.csr import AutoCSR, CSR
+
+class ICAP(Module, AutoCSR):
+    def __init__(self, platform):
+        """
+        ICAP module.
+
+        Use this module to issue the IPROG command and restart the gateware.
+        Both E2 and E3 are supported by selecting the right platform.
+        """
+        self.iprog = CSR()
+
+        ###
+
+        ICAP_WIDTH = "X32"
+        iprog_command_seq_i = [
+            0xFFFFFFFF, # 0: Dummy Word
+            0x000000BB, # 1: Bus Width Sync Word
+            0x11220044, # 2: Bus Width Detect Pattern
+            0xFFFFFFFF, # 3: Dummy Word
+            0x5599AA66, # 4: Sync Word
+            0x04000000, # 5: Type 1 NO OP 
+            0x0C400080, # 6: Write WBSTAR 
+            0x00000000, # 7: WBSTAR 
+            0x0C000180, # 8: Write CMD 
+            0x000000F0, # 9: Write IPROG 
+            0x04000000, # 10: Type 1 NO OP  
+            ]
+
+        o = Signal(32)      # 32-bit output: Configuration data output bus
+        csib = Signal()     # 1-bit input: Active-Low ICAP Enable
+        i = Signal(32)      # 32-bit input: Configuration data input bus
+        rdwrb = Signal()    # 1-bit input: Read/Write (1/0) Select input
+
+        # rising edge detection
+        edge = Signal(2)
+        self.sync += [
+            edge[1].eq(edge[0]),
+            edge[0].eq(self.iprog.re)
+        ]
+
+        fsm = FSM()
+        self.submodules += fsm
+        fsm.act(0,
+            NextValue(rdwrb, 1),
+            NextValue(csib, 1),
+            If(edge == 0b01,
+                NextState(1)
+            )
+        )
+        fsm.act(1,
+            NextValue(rdwrb, 0),
+            NextValue(csib, 1),
+            NextState(2)
+        )
+        fsm.act(2,
+            NextValue(rdwrb, 0),
+            NextValue(csib, 0),
+            NextValue(i, iprog_command_seq_i[0]),
+            NextState(3)
+        )
+        fsm.act(3,
+            NextValue(rdwrb, 0),
+            NextValue(csib, 0),
+            NextValue(i, iprog_command_seq_i[1]),
+            NextState(4)
+        )
+        fsm.act(4,
+            NextValue(rdwrb, 0),
+            NextValue(csib, 0),
+            NextValue(i, iprog_command_seq_i[2]),
+            NextState(5)
+        )
+        fsm.act(5,
+            NextValue(rdwrb, 0),
+            NextValue(csib, 0),
+            NextValue(i, iprog_command_seq_i[3]),
+            NextState(6)
+        )
+        fsm.act(6,
+            NextValue(rdwrb, 0),
+            NextValue(csib, 0),
+            NextValue(i, iprog_command_seq_i[4]),
+            NextState(7)
+        )
+        fsm.act(7,
+            NextValue(rdwrb, 0),
+            NextValue(csib, 0),
```

Comment:
No need to repeat those NextValue. And you probably do not need NextValue at all, comb assignment in the FSM probably works.